lib/AST/DiagnosticEngine.cpp

```diff
@@ -459,6 +455,20 @@ static bool typeSpellingIsAmbiguous(Type type,
   return false;
 }
 
+/// Walks the type recursivelly desugaring  types to display, but skipping
+/// `GenericTypeParamType` because we would lose association with it's original
+/// declaration and end up presenting the parameter in τ_0_0 format on
+/// diagnostic.
+static Type getAkaTypeForDisplay(Type type) {
+  auto Func = [](Type visitTy) -> Type {
+    if (isa<SugarType>(visitTy.getPointer()) &&
+        !isa<GenericTypeParamType>(visitTy.getPointer()))
+      return getAkaTypeForDisplay(visitTy->getDesugaredType());
+    return visitTy;
+  };
+  return type.transform(Func).getPointer();
```

Comment:
I think you can simplify this down to:

```
return type.transform([](Type visitType) -> Type {
  ...
});
```

since `transform` already returns a `Type`.

Reply:
Ah true, Thank you =]

Reply:
Adjusted :)